src/actors/objects/gcs.rs

```diff
@@ -0,0 +1,241 @@
+        object_id,
+        filetypes,
+        &source.files.filters,
+        source.files.layout,
+    ) {
+        let request = cache
+            .compute_memoized(FetchFileRequest {
+                scope: scope.clone(),
+                request: FetchFileInner::Gcs(source.clone(), download_path),
+                object_id: object_id.clone(),
+                threadpool: threadpool.clone(),
+            })
+            .sentry_hub_new_from_current() // new hub because of join_all
+            .map_err(|e| ArcFail(e).context(ObjectErrorKind::Caching).into())
+            .then(Ok);
+
+        requests.push(request);
+    }
+
+    Box::new(future::join_all(requests))
+}
+
+pub fn download_from_source(
+    source: Arc<GcsSourceConfig>,
+    download_path: &DownloadPath,
+) -> Box<Future<Item = Option<DownloadStream>, Error = ObjectError>> {
+    let key = {
+        let prefix = source.prefix.trim_matches(&['/'][..]);
+        if prefix.is_empty() {
+            download_path.0.clone()
+        } else {
+            format!("{}/{}", prefix, download_path.0)
+        }
+    };
+    log::debug!("Fetching from GCS: {} (from {})", &key, source.bucket);
+
+    let try_response = move || {
+        let source = source.clone();
+        let key = key.clone();
+        let url = format!(
+            "https://www.googleapis.com/upload/storage/v1/b/{}/o?name={}&uploadType={}",
+            percent_encode(source.bucket.as_bytes(), PATH_SEGMENT_ENCODE_SET),
+            percent_encode(key.as_bytes(), QUERY_ENCODE_SET),
+            "media"
+        );
+        get_token(&source.source_key)
+            .and_then(move |token| {
+                let mut builder = client::get(&url);
+                builder.header("authorization", format!("bearer {}", token.access_token));
+                builder
+                    .finish()
+                    .unwrap()
+                    .send()
+                    .map_err(|err| err.context(ObjectErrorKind::Io).into())
+            })
+            .then(move |result| match result {
+                Ok(response) => {
+                    if response.status().is_success() {
+                        log::trace!("Success hitting GCS {} (from {})", &key, source.bucket);
+                        Ok(Some(Box::new(
+                            response
+                                .payload()
+                                .map_err(|e| e.context(ObjectErrorKind::Io).into()),
+                        )
+                            as Box<dyn Stream<Item = _, Error = _>>))
+                    } else {
+                        log::trace!(
+                            "Unexpected status code from GCS {} (from {}): {}",
+                            &key,
+                            source.bucket,
+                            response.status()
+                        );
+                        Ok(None)
+                    }
+                }
+                Err(e) => {
+                    log::trace!(
```

Comment:
For the retry to work this should error on some things like 401 and wipe the token I suppose.